src/rgw/rgw_main.cc

```diff
@@ -558,6 +558,41 @@ static int process_request(RGWRados *store, RGWREST *rest, RGWRequest *req, RGWC
   dout(1) << "====== starting new request trans="  << s->trans_id.c_str() << " =====" << dendl;
   //req->log_format(s, "initializing for trans_id = %s", s->trans_id.c_str());
 
+  /* Logic for checking whether the request is token based or signature based */
+  const char* token_value = s->info.env->get("HTTP_X_AUTH_TOKEN");
+  if(token_value != NULL) {
+    // This request has a token not EC2 credentials
+    (s->auth_method).set_token_validation(true);
+    // Fill the token string even if it is blank
+    // Keystone will handle the rest
+    string value_str(token_value);
+    (s->auth_method).set_token(value_str);
+  }
+  dout(1) << "DSS INFO: token validation set to: " << (s->auth_method).get_token_validation() << dendl;
+
+  const char* amz_metadata_directive = s->info.env->get("HTTP_X_AMZ_METADATA_DIRECTIVE");
+  const char* jcs_metadata_directive = s->info.env->get("HTTP_X_JCS_METADATA_DIRECTIVE");
+  const char* amz_copy_source = s->info.env->get("HTTP_X_AMZ_COPY_SOURCE");
+  const char* jcs_copy_source = s->info.env->get("HTTP_X_JCS_COPY_SOURCE");
+
+  if(
+      (
+        (amz_metadata_directive != NULL && !strcmp(amz_metadata_directive, "COPY")) 
+        || (jcs_metadata_directive != NULL && !strcmp(jcs_metadata_directive, "COPY"))
+      ) 
+      || 
+      (amz_copy_source != NULL || jcs_copy_source != NULL)
+    ) {
+    (s->auth_method).set_copy_action(true);
```

Comment:
Make the condition AND as mentioned in the previous comment. Lets only set copy_action to true when copy source is provided.